operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/workflow/DefaultManagedWorkflow.java

```diff
@@ -10,37 +9,25 @@
 import io.fabric8.kubernetes.client.KubernetesClient;
 import io.javaoperatorsdk.operator.api.config.ControllerConfiguration;
 import io.javaoperatorsdk.operator.api.config.dependent.DependentResourceSpec;
-import io.javaoperatorsdk.operator.api.reconciler.dependent.DependentResource;
-import io.javaoperatorsdk.operator.api.reconciler.dependent.EventSourceReferencer;
-import io.javaoperatorsdk.operator.api.reconciler.dependent.NameSetter;
-
-import static io.javaoperatorsdk.operator.api.reconciler.Constants.NO_VALUE_SET;
 
 @SuppressWarnings("rawtypes")
 public class DefaultManagedWorkflow<P extends HasMetadata> implements ManagedWorkflow<P> {
+  private final DefaultWorkflow<P> workflow;
+  private boolean resolved;
+  private final List<DependentResourceSpec> orderedSpecs; // todo: remove
 
-  private final Set<String> topLevelResources;
-  private final Set<String> bottomLevelResources;
-  private final List<DependentResourceSpec> orderedSpecs;
-  private final boolean hasCleaner;
-
+  @SuppressWarnings("unchecked")
   protected DefaultManagedWorkflow(List<DependentResourceSpec> orderedSpecs, boolean hasCleaner) {
-    this.hasCleaner = hasCleaner;
-    topLevelResources = new HashSet<>(orderedSpecs.size());
-    bottomLevelResources = orderedSpecs.stream()
-        .map(DependentResourceSpec::getName)
-        .collect(Collectors.toSet());
-    this.orderedSpecs = orderedSpecs;
-    for (DependentResourceSpec<?, ?, ?> spec : orderedSpecs) {
-      // add cycle detection?
-      if (spec.getDependsOn().isEmpty()) {
-        topLevelResources.add(spec.getName());
-      } else {
-        for (String dependsOn : spec.getDependsOn()) {
-          bottomLevelResources.remove(dependsOn);
-        }
-      }
+    final var alreadyResolved = new HashMap<String, DependentResourceNode>(orderedSpecs.size());
+    for (DependentResourceSpec spec : orderedSpecs) {
+      final var node = new UnresolvedDependentResourceNode(spec);
+      alreadyResolved.put(spec.getName(), node);
+      spec.getDependsOn()
+          .forEach(depend -> node.addDependsOnRelation(alreadyResolved.get(depend)));
     }
+
+    this.workflow = new DefaultWorkflow<>(alreadyResolved.values(), false, hasCleaner);
+    this.orderedSpecs = orderedSpecs; // todo: remove
```

Comment:
I don't understand these todos in this class. Do you still plan to work on them in this PR?